### PR TITLE
dealii: add v9.6.0

### PR DESCRIPTION
**Pull request by luca-heltai**
deal.II v9.6.0

- works with kokkos+cuda through trilinos
- complex support
- opencascade support

### CODE REVIEW

**Review by jppelteret — APPROVED** [review]

**Review by masterleinad — APPROVED** [review]

**Comment by Rombur** [comment]
We need to something like:
```
 depends_on("kokkos@3.7:+cuda+cuda_lambda+cuda_constexpr+wrapper", when="@9.6:+kokkos~trilinos+cuda")
```
because deal.II 9.6 requires `cuda_constexpr` which is off by default.

**Comment by Rombur** [comment]
I am not sure if we need to change something for Trilinos

**Review by alalazo — CHANGES_REQUESTED** [review]
Failures in pipelines (`cuda_constexpr`): https://gitlab.spack.io/spack/spack/-/jobs/12191568

**Comment by masterleinad** [comment]
You need to fix the formatting
```
--- var/spack/repos/builtin/packages/dealii/package.py	2024-08-27 14:50:05.049501+00:00
+++ var/spack/repos/builtin/packages/dealii/package.py	2024-08-27 14:50:19.914640+00:00
@@ -199,11 +199,13 @@
     # TODO: next line fixes concretization with petsc
     depends_on("hdf5+mpi+hl+fortran", when="+hdf5+mpi+petsc")
     depends_on("hdf5+mpi+hl", when="+hdf5+mpi~petsc")
     depends_on("kokkos@3.7:", when="@9.5:+kokkos~trilinos")
     depends_on("kokkos@3.7:+cuda+cuda_lambda+wrapper", when="@9.5:9.5.99+kokkos~trilinos+cuda")
-    depends_on("kokkos@3.7:+cuda+cuda_lambda+cuda_constexpr+wrapper", when="@9.6:+kokkos~trilinos+cuda")
+    depends_on(
+        "kokkos@3.7:+cuda+cuda_lambda+cuda_constexpr+wrapper", when="@9.6:+kokkos~trilinos+cuda"
+    )
     # TODO: concretizer bug. The two lines mimic what comes from PETSc
     # but we should not need it
     depends_on("metis@5:+int64", when="+metis+int64")
     depends_on("metis@5:~int64", when="+metis~int64")
     depends_on("muparser", when="+muparser")
```

**Comment by luca-heltai** [comment]
@masterleinad, as soon as #45812 is merged, we can add the required change also here.

**Comment by adamqc** [comment]
@luca-heltai I need the `complex` variant to compile with complex value support. Could you please add it to the package? Please see #46539 for reference. Thanks!

**Comment by luca-heltai** [comment]
The failing CI is unrelated to this PR (vtk fails building). Can we merge this?

File: var/spack/repos/builtin/packages/dealii/package.py
Comment (masterleinad) [inline var/spack/repos/builtin/packages/dealii/package.py:210]: Don't we also need something like
```
depends_on("trilinos+cuda+cuda_constexpr", when "@9.6:+trilinos+cuda")
```
?

**Comment by eugeneswalker** [comment]
Does `dealii@9.6.0` support `cuda@12`?

**Comment by masterleinad** [comment]
> Does `dealii@9.6.0` support `cuda@12`?

Not via `DEAL_II_WITH_CUDA` but only through `Kokkos`.

**Comment by luca-heltai** [comment]
Can we get this merged? I think the E4S build can disable vtk, if they want this to build. I'm not sure what the problem with vtk is, but it's not related to this PR.

**Comment by eugeneswalker** [comment]
> Can we get this merged? I think the E4S build can disable vtk, if they want this to build. I'm not sure what the problem with vtk is, but it's not related to this PR.

Just pushed commit to set `dealii ~vtk` in CI stacks.

**Comment by eugeneswalker** [comment]
More tweaks to E4S CI stacks here:
* https://github.com/spack/spack/pull/45554/commits/86a0ed8f278ffc368987639e90d3b3a8fb318232

1. E4S Neoverse V1: disable dealii due to build errors
* [dealii@9.6.0 /7kn4cfi %gcc@11.4.0 arch=linux-ubuntu22.04-neoverse_v1 E4S ARM Neoverse V1 ](https://gitlab.spack.io/spack/spack/-/jobs/13385628)
```
...
  >> 2904    /tmp/root/spack-stage/spack-stage-dealii-9.6.0-7kn4cfiil6krffy6c4l
             vjdrmztyhju3w/spack-src/include/deal.II/base/vectorization.h(1553)
             : error: identifier "vminq_f32" is undefined
     2905    
     2906    28 errors detected in the compilation of "/tmp/root/spack-stage/sp
             ack-stage-dealii-9.6.0-7kn4cfiil6krffy6c4lvjdrmztyhju3w/spack-src/
             source/dofs/dof_tools_sparsity.cc".
     2907    28 errors detected in the compilation of "/tmp/root/spack-stage/sp
             ack-stage-dealii-9.6.0-7kn4cfiil6krffy6c4lvjdrmztyhju3w/spack-src/
             source/dofs/dof_tools_constraints.cc".
  >> 2908    make[2]: *** [source/dofs/CMakeFiles/object_dofs_release.dir/build
             .make:247: source/dofs/CMakeFiles/object_dofs_release.dir/dof_tool
             s_sparsity.cc.o] Error 1
...
```

2. E4S OneAPI: Set `dealii ~vtk` due to issue with dependencies not being buildable by OneAPI
* [ taskflow@3.7.0 /rm3sdb3 %oneapi@2024.2.1 arch=linux-ubuntu22.04-x86_64_v3 E4S OneAPI ](https://gitlab.spack.io/spack/spack/-/jobs/13385607)
```
  >> 10    CMake Error at CMakeLists.txt:81 (message):
     11    
     12    
     13      Taskflow currently supports the following compilers:
     14    
     15        - g++ v7.0 or above
     16        - clang++ v6.0 or above
```

**Comment by balay** [comment]
https://gitlab.spack.io/spack/spack/-/jobs/13482234
```

PLEASE append the compiler options "-save-temps -v", rebuild the application to get the full command which is failing and submit a bug report to https://software.intel.com/en-us/support/priority-support which includes the failing command, input files for the command and the crash backtrace (if any).
Stack dump:
0.      Program arguments: /opt/intel/oneapi/compiler/2024.2/bin/compiler/clang++ @/tmp/icpx0013442874BhV7Ra/icpxargkyxtt2
1.      <eof> parser at end of file
2.      Optimizer
```
>  -qopenmp-simd -std=c++17 -march=x86-64-v3 -mtune=generic -O2

perhaps check if `cflags=-O0 cxxflags-O0` will work around this compiler  failure? [or somehow disable ` -qopenmp-simd ` and see if that works?]

File: share/spack/gitlab/cloud_pipelines/stacks/e4s-oneapi/spack.yaml
Comment (balay) [inline share/spack/gitlab/cloud_pipelines/stacks/e4s-oneapi/spack.yaml]:
```suggestion
  - dealii ~vtk ~taskflow cflags=-O0 cxxflags=-O0 # +vtk +taskflow: ^taskflow: CMake Error at CMakeLists.txt:81 (message): Taskflow currently supports the following compilers: g++ v7.0 or above, clang++ v6.0 or above; use -O0 to work around compiler failure
```

**Comment by balay** [comment]
The CI jobs are clean now!

I think this PR is ready for merge!

[assuming  my above CI change is acceptable (cc: @eugeneswalker ) and @alalazo can approve an clear the block on this PR]

**Comment by eugeneswalker** [comment]
> I think this PR is ready for merge!
> 
> [assuming my above CI change is acceptable (cc: @eugeneswalker ) and @alalazo can approve an clear the block on this PR]

Looks good to me! Thanks for doing that!

**Review by eugeneswalker — APPROVED** [review]

**Review by alalazo — APPROVED** [review]

**Pull request merged**
